File: archive/java/BeeBattle.java
/*
	Final Project -Game
	Todd Burnop
	David Rawson Couzelis
	-Function 12 Software
	Start::11.20.2000

	
*/


import java.io.*;
import java.awt.*;
import java.awt.event.*;
import java.awt.image.*;
import java.util.*;
import java.lang.*;



class MainSystem 
{

	public static void main (String args[])  
	{
		
		GameWorld game = new GameWorld ();	
		
		game.playGameWorld ();

	}
}





//A generic Opposing Object...
class OO 
{

	boolean alive;                      //If the OO is still alive...
	protected int x_pos, y_pos;         //OO's current position...
	protected int change_x, change_y;   //The change in position each game update...
	protected int direction;            //0 for left, 1 for right...
	
	protected Image pic;                //Stores the picture!...
	
	
	
	
	
	public Image getPic ()  {  return (pic);  }
	public int getXPos ()  {  return (x_pos);  }
	public int getYPos ()  {  return (y_pos);  }
	
	
	//Construct an OO!!...
	OO (String nameSent, int start_x, int start_y)
	{
		alive = true;
		direction = 1;

		pic = Toolkit.getDefaultToolkit().getImage(nameSent);
			
		x_pos = start_x;
		y_pos = start_y;
	
	}


	//Update each OO each game loop...
	public OO update ()
	{
	
		if (alive)
		{
		
			//Move OO in the correct direction...
			if (direction == 0)
				x_pos -= 2;
			if (direction == 1)
				x_pos += 2;
	
			//Check for the hitting the edge of the screen!...	
			if (x_pos >= 640-50)
			{
				x_pos = 640-50;
				direction = 0;
			}
			if (x_pos <= 0)
			{
				x_pos = 0;
				direction = 1;
			}
			
			return (this);
		}
		else  return (null);
		
	}
	
	
	
	public boolean shoot (int CCpos)
	{
		//Code here whenever you want the OOs to shoot...
		
		//Randomly run through numbers, to see if the OO will shoot...
		//One of the simplest AIs I can think of  =)  ...
		if ((int)(java.lang.Math.random () * 500) == 0)
				return (true);
		
		
		//Put a little bit more often random generator here...
		//If true, and are facing the CC, shoot...
		if (CCpos == x_pos && (int)(java.lang.Math.random () * 4) == 0)
			return (true);
		
		
		else  return (false);
	
	}
	

}



class ATK
{


	int x_pos, y_pos;          //Current ATK position...
	int change_x, change_y;    //The change in position each game update...
	int type;                  //1 for CC, 2 for OO...
	
	Image pic;                 //The picture of the ATK...

	


	public Image getPic ()  {  return (pic);  }
	public int getXPos ()  {  return (x_pos);  }
	public int getYPos ()  {  return (y_pos);  }

	
	ATK (String nameSent, int x, int y, int typeSent)
	{
		x_pos = x;  y_pos = y;

		type = typeSent;
		if (type == 1)  {  change_x = 0;  change_y = -15;  }
		else  if (type == 2)  {  change_x = 0;  change_y = 5;  }
		
		pic = Toolkit.getDefaultToolkit().getImage(nameSent);
	
	}
	
	
	public ATK update ()
	{
		x_pos += change_x;
		y_pos += change_y;

		//Check if the ATK is off the screen...
		if (x_pos<=0-20 || x_pos>=640 || y_pos<=0-20 || y_pos>=480)
			return (null);  
		else  return (this);	
			
	}


}




class GameWindow extends Frame 
{

	//NOTE::  Later, add another screen component, draw to it, then display.  Buffer!!...


	//NEW::  Try the idea that, since everything will take place in the GameWindow,
	//everything should be created in that...
	//NOTE::  Don't forget to validate after every redraw!!...

	protected static CC cc1;        //The good guy!...
	protected static OO[] oo;       //The list of OOs...
	protected static ATK[] CCatk;   //The list of CC ATKs...
	protected static ATK[] OOatk;   //The list of OO ATKs...
	
	
	Image background;               //Stores the background pic...
	Image title;
	char inkey;


	public boolean keepPlaying;

	protected Image buffer;         //Draw to this, then display buffer...


	//Override Update to minimize flicker...
	//(Don't ask me why!)...
	public void update (Graphics g) 
	{
		paint (g);
	}


	
	GameWindow ()
	{
	
	
		//Initialize the GameWorld objects...
		cc1 = new CC (5);    
	
		oo = new OO[40];
		
		
		int count = 0;
		//Initialize the OOs however you want!...
		for (int i=0; i<3; i+=1)
			for (int j=0; j<5; j+=1)
			{
				oo[count] = new OO ("oo1.gif", j*60+20,i*60+20);
				count += 1;
			}
	
		CCatk = new ATK[5];
		OOatk = new ATK[20];
		for (int i=0; i<5; i+=1)
			CCatk[i] = null;
		for (int i=0; i<20; i+=1)
			OOatk[i] = null;
	
		background = Toolkit.getDefaultToolkit().getImage("bg.gif");
		title = Toolkit.getDefaultToolkit().getImage("title.gif");
		inkey = '~';
	
		keepPlaying = true;

	
		//Ha!  WindowListeners ain't so hard...
		addWindowListener (new WindowAdapter ()
			{
			public  void windowClosing (WindowEvent e)
				{
					System.exit (0);
				}
			}  );


		//For key input...
		addKeyListener (new KeyAdapter ()
			{
			public void keyPressed (KeyEvent e)
				{
					inkey = e.getKeyChar();
				}
			public void keyReleased (KeyEvent e)
				{
					inkey = '~';
				}
			}  );
			
	
	}
	
	
	public void paint (Graphics g)
	{
	
		//Ready the buffer and ready the Graphics...
		buffer = createImage (640,480);
		Graphics bg = buffer.getGraphics();	
		//bg = g;
		
		bg.drawImage (background, 0,0, null);		
		bg.drawImage (cc1.getPic (), cc1.getXPos(),cc1.getYPos(), null);
		
		for (int i=0; i<40; i+=1)
			if (oo[i] != null)
				bg.drawImage (oo[i].getPic (), oo[i].getXPos(),oo[i].getYPos(), null);
		
		for (int i=0; i<5; i+=1)
			if (CCatk[i] != null)
				bg.drawImage (CCatk[i].getPic (), CCatk[i].getXPos(),CCatk[i].getYPos(), null);
		for (int i=0; i<20; i+=1)
			if (OOatk[i] != null)
				bg.drawImage (OOatk[i].getPic (), OOatk[i].getXPos(),OOatk[i].getYPos(), null);
		
		//Finalize the draw!...
		g.drawImage (buffer, 0,0, null);
		
		
	}


	

	public void updateEverything ()
	{
		
		//Update the CC...
		cc1.update (inkey);
		//No more than 5 CC ATKs on screen...
		if (cc1.shoot (inkey))
		{
			int j;
			for (j=0; CCatk[j]!=null && j<4; j+=1)
				;
			CCatk[j] = new ATK ("ccatk.gif", cc1.getXPos()+15,cc1.getYPos()+15, 1);	
			inkey = '~';

		}
	
	
		//Update the OOs...
		for (int i=0; i<40; i+=1)
			if (oo[i] != null)
			{
				
				if (oo[i].shoot(cc1.getXPos()))
				{
					int j;
					for (j=0; OOatk[j]!=null && j<20; j+=1)
						;
					OOatk[j] = new ATK ("ooatk.gif", oo[i].getXPos()+15,oo[i].getYPos()+15, 2);	
				
				}
				
				
				oo[i] = oo[i].update ();	
				
			}
	
		
		//Update the CC ATKs...
		for (int i=0; i<5; i+=1)
		{	
			
			if (CCatk[i] != null)
			{
				
				//Check for OO hit!!...
				for (int j=0; j<40; j+=1)
					if (oo[j]!=null)
						if (CCatk[i]!=null)
							if (CCatk[i].getXPos()>oo[j].getXPos() && (CCatk[i].getXPos()+20)<(oo[j].getXPos()+50))
								if (CCatk[i].getYPos()>oo[j].getYPos() && (CCatk[i].getYPos()+20)<(oo[j].getYPos()+50))
								{
									//Kill the ATK and the hit OO!!...
									CCatk[i] = null;
									oo[j] = null;
								}
						
				//After all that checking, if the CC ATK is still there, update it!...
				if (CCatk[i] != null)
					CCatk[i] = CCatk[i].update ();

			}
			
		}
		
		//Update the OO ATKs...
		for (int i=0; i<20; i+=1)
		{
			if (OOatk[i] != null)
			{
			
				//YOU LEFT OFF HERE!!			
			
				if (OOatk[i] != null)
					OOatk[i] = OOatk[i].update();
				
				
			}	


		}


		
	}


}




//The main game world...
class GameWorld 
{

	
	//Create the game window here, and construct it...
	GameWindow screen;
	long timer;	
	
	
	public GameWorld ()
	{
	
		//Initialize the window!...
		screen = new GameWindow ();
		
		screen.setTitle ("Bee Battle!!");
		screen.setSize (640,480);
		screen.setResizable (false);
		screen.setLocation (30,30);
		screen.setVisible (true);
		
		timer = System.currentTimeMillis ();	
		
	}




	//Call this to get the game started.  It will finish when the game is over!...
	public void playGameWorld () 
	{
						
		while (screen.keepPlaying)
		{
			//Update the game...
			//Send the current inkey...
			screen.updateEverything ();
			
			//Draw Screen!!...
			screen.repaint ();
			//screen.validate();

			//Wait!!...			
			//(try to replace this with a nicer, friendlier, faster wait)...
			while (System.currentTimeMillis() - timer < 16);
			timer = System.currentTimeMillis ();	
			
		
		}
	
	
	}


}






//A generic Controllable Character...
class CC 
{

	protected int lives;               //Amount of remaining lives...
	protected int x_pos, y_pos;        //The position of the CC...
	
	protected Image pic1, pic2;        //Stores the two pictures to be animated...
	protected int frame;               //Which tile to display now...


	//Construct a CC!!...
	CC (int livesSent)
	{
		lives = livesSent;
		x_pos = 640/2;     //Start at the center of the screen for now...
		y_pos = 480-100;   //Near the bottom of the screen...
		
		pic1 = Toolkit.getDefaultToolkit().getImage("cc1.gif");
		pic2 = Toolkit.getDefaultToolkit().getImage("cc2.gif");
		frame = 1;

	}


	//Ask the CC if it's ready to shoot!...
	public boolean shoot (char inkey)
	{
		if (inkey == 'z' || inkey == ' ')
			return (true);
		else  return (false);

	}


 	public void update (char inkey)  
	{  

		//Update the pos (keyboard input)...
		if (inkey == 'j')
			x_pos -= 15;
		else  if (inkey == 'l')
			x_pos += 15;
		
		
		//Check out of bounds position!...
		if (x_pos > 640-50)
			x_pos = 640-50;
		else  if (x_pos < 0)
			x_pos = 0;
		if (y_pos > 480-50)
			y_pos = 48050;
		else  if (y_pos < 0)
			y_pos = 0;		
		
	
	}


	public Image getPic ()  
	{  
		if (frame == 1)  {  frame=2;  return (pic1);  }
		else  {  frame=1;  return (pic2);  }
	}
	public int getXPos ()  {  return (x_pos);  }
	public int getYPos ()  {  return (y_pos);  }



	//When the CC is hit, kill him!...
	//Return true when there are no more extra lives...
	public boolean kill ()
	{
		x_pos = 640/2;
		lives -= 1;
		if (lives < 1)
			return (true);
		else
			return (false);
	}


}

File: archive/objc/beebattle.m
/**
 * Copyright 2009 David Couzelis
 * 
 * This file is part of "Bee Battle".
 * 
 * "Bee Battle" is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * "Bee Battle" is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with "Bee Battle".  If not, see <http://www.gnu.org/licenses/>.
 */
/**
 * Final Project Game
 * Function 12 Software
 * David Rawson Couzelis
 * Todd Burnop
 * Start 2000/11/20
 */
#import <objc/Object.h>
#import "allegro.h"


#define GAME_TICKER 60

#define SCREEN_WIDTH 640
#define SCREEN_HEIGHT 480
#define COLOR_DEPTH 24

#define BLACK (makecol(0, 0, 0))

#define MAX_NUM_OF_SPRITES 256
#define MAX_NUM_OF_CC_ATK 2


volatile int timer = 0;


int random_number(int low, int high) {
  return (rand() % (high - low + 1)) + low;
}


@interface ATK : Object {
  
  int x_pos, y_pos;          // Current ATK position...
  int x_speed, y_speed;
  
  BITMAP *pic;               // The picture of the ATK...
  int type;                  // 1 for CC, 2 for OO...
  
}

- initWithImage: (char *) nameSent andX: (int) start_x andY: (int) start_y andType: (int) typeSent;
- update;
- (BITMAP *) getPic;
- (int) getXPos;
- (int) getYPos;

@end


@implementation ATK


- init {
  self = [super init];
  if (self) {
    x_pos = 0;
    y_pos = 0;
    x_speed = 0;
    y_speed = 0;
    type = 0;
    pic = NULL;
  }
  return self;
}


- initWithImage: (char *) nameSent andX: (int) start_x andY: (int) start_y andType: (int) typeSent {
  
  PALLETE my_pallete;
  
  self = [self init];
  
  if (self) {
    
    x_pos = start_x;
    y_pos = start_y;
    type = typeSent;
    
    if (type == 1) {
      y_speed = -6;
    } else if (type == 2) {
      y_speed = 2;
    }
    
    pic = load_bitmap(nameSent, my_pallete);
    set_palette(my_pallete);
    
  }
  
  return self;
  
}


- free {
  destroy_bitmap(pic);
  return [super free];
}


- update {
  
  x_pos += x_speed;
  y_pos += y_speed;
  
  // Check if the ATK is off the screen...
  if (x_pos <= -[self getPic]->w || x_pos >= SCREEN_WIDTH || y_pos <= -[self getPic]->h || y_pos >= SCREEN_HEIGHT) {
    return nil;
  }
  
  return self;
  
}


- (BITMAP *) getPic {
  return pic;
}


- (int) getXPos {
  return x_pos;
}


- (int) getYPos {
  return y_pos;
}


@end


// A generic Controllable Character...
@interface CC : Object {
  
  int lives;               // Amount of remaining lives...
  int x_pos, y_pos;        // The position of the CC...
  
  BITMAP *pic1;            // Stores the two pictures to be animated...
  BITMAP *pic2;
  int frame;               // Which tile to display now...
  
  int keyCount;            // Counts how long the shoot button has been held...
  
}

- initWithLives: (int) livesSent;
- (BOOL) shoot;
- update;
- (BITMAP *) getPic;
- (int) getXPos;
- (int) getYPos;
- (BOOL) kill;

@end


@implementation CC


- init {
  self = [super init];
  if (self) {
    lives = 0;
    x_pos = 0;
    y_pos = 0;
    pic1 = NULL;
    pic2 = NULL;
    frame = 0;
    keyCount = 0;
  }
  return self;
}


// Construct a CC!!...
- initWithLives: (int) livesSent {
  PALLETE my_pallete;
  self = [self init];
  if (self) {
    lives = livesSent;
    pic1 = load_bitmap("bee1.bmp", my_pallete);
    pic2 = load_bitmap("bee2.bmp", my_pallete);
    set_palette(my_pallete);
    frame = 1;
    x_pos = (SCREEN_WIDTH / 2) - ([self getPic]->w / 2);   // Start at the center of the screen for now...
    y_pos = SCREEN_HEIGHT - ([self getPic]->h * 2);        // Near the bottom of the screen...
  }
  return self;
}


- free {
  destroy_bitmap(pic1);
  destroy_bitmap(pic2);
  return [super free];
}


// Ask the CC if it's ready to shoot!...
- (BOOL) shoot {
  if (key[KEY_SPACE]) {
    if (keyCount == 0) {
      keyCount = 20;
      return YES;
    } else {
      keyCount--;
      return NO;
    }
  }
  keyCount = 0;
  return NO;
}


- update {
  
  //Update the pos (keyboard input)...
  if (key[KEY_LEFT]) {
    x_pos -= 4;
  }
  
  if (key[KEY_RIGHT]) {
    x_pos += 4;
  }
  
  // Check out of bounds position!...
  if (x_pos > SCREEN_WIDTH - [self getPic]->w) {
    x_pos = SCREEN_WIDTH - [self getPic]->w;
  } else if (x_pos < 0) {
    x_pos = 0;
  }
  
  if (y_pos > SCREEN_HEIGHT - [self getPic]->h) {
    y_pos = SCREEN_HEIGHT - [self getPic]->h;
  } else if (y_pos < 0) {
    y_pos = 0;
  }
  
  return self;
  
}


- (BITMAP *) getPic {
  if (frame == 1) {
    frame = 2;
    return pic1;
  } else {
    frame = 1;
    return pic2;
  }
}

- (int) getXPos {
  return x_pos;
}


- (int) getYPos {
  return y_pos;
}


// When the CC is hit, kill him!...
// Return true when there are no more extra lives...
- (BOOL) kill {
  x_pos = (SCREEN_WIDTH / 2) - ([self getPic]->w / 2);     // Start at the center of the screen for now...
  y_pos = SCREEN_HEIGHT - ([self getPic]->h * 2);   // Near the bottom of the screen...
  lives--;
  if (lives == 0) {
    return YES;
  }
  return NO;
}


@end


// A generic Opposing Object...
@interface OO : Object {
  
  int x_pos, y_pos;         // OO's current position...
  int x_speed, y_speed;
  
  int direction;            // 0 for left, 1 for right...
  
  BITMAP *pic;              // Stores the picture!...
  
}

- initWithImage: (char *) imageFilename andX: (int) start_x andY: (int) start_y;
- update;
- (BOOL) shoot: (int) CCpos;
- (BITMAP *) getPic;
- (int) getXPos;
- (int) getYPos;

@end


@implementation OO


// Construct an OO!!...
- init {
  self = [super init];
  if (self) {
    direction = 1;
    x_pos = 0;
    y_pos = 0;
    x_speed = 0;
    y_speed = 0;
    pic = NULL;
  }
  return self;
}

  
- initWithImage: (char *) nameSent andX: (int) start_x andY: (int) start_y {
  PALLETE my_pallete;
  self = [self init];
  if (self) {
    pic = load_bitmap(nameSent, my_pallete);
    set_palette(my_pallete);
    x_pos = start_x;
    y_pos = start_y;
    x_speed = 1;
  }
  return self;
}


- free {
  destroy_bitmap(pic);
  return [super free];
}


// Update each OO each game loop...
- update {
  
  // Move the OO in the correct direction...
  if (direction == 0) {
    x_pos -= x_speed;
  } else {
    x_pos += x_speed;
  }
  
  // Check for the hitting the edge of the screen!...	
  if (x_pos >= SCREEN_WIDTH - [self getPic]->w) {
    x_pos = SCREEN_WIDTH - [self getPic]->w;
    direction = 0;
  }
  
  if (x_pos <= 0) {
    x_pos = 0;
    direction = 1;
  }
  
  return (self);
  
}


- (BOOL) shoot: (int) CCpos {
  
  // Code here whenever you want the OOs to shoot...
  
  // Randomly run through numbers, to see if the OO will shoot...
  // One of the simplest AIs I can think of  =)  ...
  if (random_number(0, 500) == 0) {
    return YES;
  }
  
  // Put a little bit more often random generator here...
  // If true, and are facing the CC, shoot...
  if (CCpos == x_pos && random_number(0, 4) == 0) {
    return YES;
  }
  
  return NO;
  
}


- (BITMAP *) getPic {
  return pic;
}


- (int) getXPos {
  return x_pos;
}


- (int) getYPos {
  return y_pos;
}


@end


@interface World : Object {
  
  // NOTE:: Later, add another screen component, draw to it, then display. Buffer!!...
  
  CC *cc1;                          //The good guy!...
  OO *oo[MAX_NUM_OF_SPRITES];       //The list of OOs...
  ATK *CCatk[MAX_NUM_OF_SPRITES];   //The list of CC ATKs...
  ATK *OOatk[MAX_NUM_OF_SPRITES];   //The list of OO ATKs...
  
  BITMAP *background;               //Stores the background pic...
  BITMAP *title;
  
  BOOL keepPlaying;
  
  BITMAP *buffer;                   //Draw to this, then display buffer...
  
}

- update;
- draw;
- (BOOL) keepPlaying;

@end


@implementation World


- init {
  
  PALLETE my_pallete;
  int count;
  int i, j;
  
  self = [super init];
  
  if (self) {
    
    cc1 = nil;
    
    for (i = 0; i < MAX_NUM_OF_SPRITES; i++) {
      oo[i] = nil;
    }
    
    for (i = 0; i < MAX_NUM_OF_SPRITES; i++) {
      CCatk[i] = nil;
    }
    
    for (i = 0; i < MAX_NUM_OF_SPRITES; i++) {
      OOatk[i] = nil;
    }
    
    // Initialize the GameWorld objects...
    cc1 = [[CC alloc] initWithLives: 5];
    
    //oo = new OO[40];
    
    count = 0;
    
    // Initialize the OOs however you want!...
    for (i = 0; i < 3; i++) {
      for (j = 0; j < 5; j++) {
        oo[count] = [[OO alloc] initWithImage: "flower.bmp" andX: j * 60 + 20 andY: i * 60 + 20];
        count++;
      }
    }
    
    background = load_bitmap("background.bmp", my_pallete);
    title = load_bitmap("title.bmp", my_pallete);
    set_palette(my_pallete);
    
    keepPlaying = YES;
    
    buffer = create_bitmap(SCREEN_WIDTH, SCREEN_HEIGHT);
    clear_to_color(buffer, BLACK);
    
  }
  
  return self;
  
}


- free {
  
  int i;
  
  [cc1 free];
  
  for (i = 0; i < MAX_NUM_OF_SPRITES; i++) {
    [oo[i] free];
  }
  
  for (i = 0; i < MAX_NUM_OF_SPRITES; i++) {
    [CCatk[i] free];
  }
  
  for (i = 0; i < MAX_NUM_OF_SPRITES; i++) {
    [OOatk[i] free];
  }
  
  destroy_bitmap(background);
  destroy_bitmap(title);
  
  destroy_bitmap(buffer);
  
  return self;
  
}

  
- update {
  
  int i, j;
  
  // Update the CC...
  [cc1 update];
  
  if ([cc1 shoot]) {
    i = 0;
    while (i < MAX_NUM_OF_CC_ATK && CCatk[i] != nil) {
      i++;
    }
    if (i < MAX_NUM_OF_CC_ATK) {
      CCatk[i] = [[ATK alloc] initWithImage: "stinger.bmp" andX: [cc1 getXPos] + 15 andY: [cc1 getYPos] + 15 andType: 1];
    }
  }	
  
  // Update the OOs...
  for (i = 0; i < MAX_NUM_OF_SPRITES; i++) {
    if (oo[i] != nil) {
      
      if ([oo[i] shoot: [cc1 getXPos]]) {
        j = 0;
        while (j < MAX_NUM_OF_SPRITES && OOatk[j] != nil) {
          j++;
        }
        if (i < MAX_NUM_OF_SPRITES) {
          OOatk[j] = [[ATK alloc] initWithImage: "pollen.bmp" andX: [oo[i] getXPos] + 15 andY: [oo[i] getYPos] + 15 andType: 2];
        }
      }
      
      [oo[i] update];
      
    }
  }
  
  // Update the CC ATKs...
  for (i = 0; i < MAX_NUM_OF_SPRITES; i++) {
    
    if (CCatk[i] != nil) {
      
      // Check for OO hit!!...
      for (j = 0; j < MAX_NUM_OF_SPRITES; j++) {
        if (oo[j] != nil) {
          if (CCatk[i] != nil) {
            if ([CCatk[i] getXPos] > [oo[j] getXPos] && [CCatk[i] getXPos] + [CCatk[i] getPic]->w < [oo[j] getXPos] + [oo[j] getPic]->w) {
              if ([CCatk[i] getYPos] > [oo[j] getYPos] && [CCatk[i] getYPos] + [CCatk[i] getPic]->h < [oo[j] getYPos] + [oo[j] getPic]->h) {
                // Kill the ATK and the hit OO!!...
                [CCatk[i] free];
                CCatk[i] = nil;
                [oo[j] free];
                oo[j] = nil;
              }
            }
          }
        }
      }
      
      // After all that checking, if the CC ATK is still there, update it!...
      if (CCatk[i] != nil) {
        [CCatk[i] update];
        if ([CCatk[i] getYPos] < 0) {
          [CCatk[i] free];
          CCatk[i] = nil;
        }
      }
      
    }
    
  }
  
  // Update the OO ATKs...
  for (i = 0; i < MAX_NUM_OF_SPRITES; i++) {
    if (OOatk[i] != nil) {
      
      // YOU LEFT OFF HERE!!
      //if (OOatk[i] != nil) {
        [OOatk[i] update];
        if ([OOatk[i] getYPos] > SCREEN_HEIGHT) {
          [OOatk[i] free];
          OOatk[i] = nil;
        }
      //}
      
    }
  }
  
  return self;
  
}


- draw {
  
  int i;
  
  // Ready the buffer and ready the Graphics...
  //buffer = createImage (640,480);
  //Graphics bg = buffer.getGraphics();	
  
  draw_sprite(buffer, background, 0, 0);
  draw_sprite(buffer, [cc1 getPic], [cc1 getXPos], [cc1 getYPos]);
  
  for (i = 0; i < MAX_NUM_OF_SPRITES; i++) {
    if (oo[i] != nil) {
      draw_sprite(buffer, [oo[i] getPic], [oo[i] getXPos], [oo[i] getYPos]);
    }
  }
  
  for (i = 0; i < MAX_NUM_OF_SPRITES; i++) {
    if (CCatk[i] != nil) {
      draw_sprite(buffer, [CCatk[i] getPic], [CCatk[i] getXPos], [CCatk[i] getYPos]);
    }
  }
  
  for (i = 0; i < MAX_NUM_OF_SPRITES; i++) {
    if (OOatk[i] != nil) {
      draw_sprite(buffer, [OOatk[i] getPic], [OOatk[i] getXPos], [OOatk[i] getYPos]);
    }
  }
  
  // Finalize the draw!...
  vsync();
  blit(buffer, screen, 0, 0, 0, 0, buffer->w, buffer->h);
  
  return self;
  
}


- (BOOL) keepPlaying {
  return keepPlaying;
}


@end


// The main game world...
@interface Game : Object {
  // Create the game window here, and construct it...
  World *world;
  // YOU LEFT OFF HERE!!
  //long timer;
}

- play;

@end


@implementation Game


- init {
  self = [super init];
  if (self) {
    // Initialize the window!...
    world = [[World alloc] init];
    //screen.setTitle ("Bee Battle!!");
    //screen.setSize (640,480);
    //screen.setResizable (false);
    //screen.setLocation (30,30);
    //screen.setVisible (true);
    // YOU LEFT OFF HERE!!
    //timer = System.currentTimeMillis ();
  }
  return self;
}


- free {
  [world free];
  return [super free];
}


// Call this to get the game started. It will finish when the game is over!...
- play {
  
  int timemark;
  
  timer = 0;
  
  while (!key[KEY_ESC] && [world keepPlaying]) {
    
    while (timer == 0) {
      rest(100 / GAME_TICKER);
    }
    
    while (timer > 0) {
      
      timemark = timer;
      
      // Update the game...
      [world update];
      
      timer--;
      
      if (timemark <= timer) {
        break;
      }
      
    }
    
    // Draw Screen!!...
    [world draw];
    
    // Wait!!...
    // (try to replace this with a nicer, friendlier, faster wait)...
    // Fixed: Please see the use of the timer above.
    //while (System.currentTimeMillis() - timer < 16);
    //timer = System.currentTimeMillis ();	
    
  }
  
  return self;
  
}


@end


/**
 * To keep the game running at the correct frames per second
 */
void do_timer(void) {
  timer++;
} END_OF_FUNCTION (do_timer);


int main() {
  
  Game *game;
  
  allegro_init();
  
  install_timer();
  
  LOCK_VARIABLE(timer);
  LOCK_FUNCTION(do_timer);
  install_int_ex(do_timer, BPS_TO_TIMER(GAME_TICKER));
  
  srand(time(NULL));
  
  install_keyboard();
  
  set_color_depth(COLOR_DEPTH);
  set_gfx_mode(GFX_AUTODETECT_WINDOWED, SCREEN_WIDTH, SCREEN_HEIGHT, 0, 0);
  clear_to_color(screen, makecol(0, 0, 0));
  
  game = [[Game alloc] init];
  [game play];
  [game free];
  
  return 0;
  
}
END_OF_MAIN()
